### ISSUE TITLE
Plannable Import: ID Interpolation

### ISSUE THREAD

**Issue by bondsb** [hashicorp/terraform#33228]
### Terraform Version

```shell
Terraform v1.5.0-beta1
on darwin_amd64
```

### Use Cases

Interpolation of the import block `id` enables more complex import scenarios, allowing the user to specify identifiers through variables or other known values at plan time.

This feature enables easier import solutions for

- Pipelines
- Terraform Cloud
- Self-Service Solutions

It also provides a mechanism for automatic import, or "import-if-not-exists".

### Attempted Solutions

The **plannable import** feature currently limits the value of the import block `id` to a constant string value.  The following code snippet is not allowed:

```tf
resource "aws_instance" "web" {
    ...
}

import {
    to = aws_instance.web
    id = var.web_instance_id  # FAILS
}
```

### Proposal

This proposal is an extension of the **plannable import** feature which is planned for inclusion in Terraform 1.5.

In this feature, the identifier may be specified by any interpolation that can be resolved at plan time.  In this example (the same as the above example), we are now allowed to use a variable for an AWS instance ID:

```tf
resource "aws_instance" "web" {
    ...
}

import {
    to = aws_instance.web
    id = var.web_instance_id  # allowed now
}
```

The import block is ignored if `id` is `null` and a new resource will be created.  Conditional import with variables makes it easier to import resources using pipelines or Terraform Cloud.

---
###

For this scenario, we would like to import a web server if it exists or create a new one if it doesn't.  The cloud provider creates a new ID every time we rebuild our infrastructure.  But we know that the server will always be named "webserver1".

Let's use use a data source to check whether "webserver1" exists:

```tf
data "aws_instance" "existing_server" {
  filter {
    name   = "tag:Name"
    values = ["webserver1"]
  }
}
```

and import it only if it does:
```tf
resource "aws_instance" "web" {
    ...
}

import {
    to = aws_instance.web
    id = try(data.aws_instance.existing_server.id, null)  # if the server doesn't exist, create a new one
}
```

### References

- https://github.com/hashicorp/terraform/pull/33085#issuecomment-1555187415
- #27959
- https://github.com/hashicorp/terraform/pull/31910#issuecomment-1263171016
- #33216
- #26364

**Comment by crw** [comment]
Thanks for submitting this request!

**Comment by ilmax** [comment]
On top of what @bondsb  suggested above, I would like to use import blocks **NOT** only in the root module.
When you have a complex infrastructure, it's a common practice to break it down in separate modules to make it more manageable. 
In this scenario it would be preferred to be able to put the import blocks alongside the resource declarations, this will also allow/simplify using variables (or interpolation) in the import id that may not be available/in scope in the root module.

Is there a reason why import is only allowed on root modules or is it a just an initial limitation?

**Comment by omarismail** [comment]
Hey @ilmax, thanks for your input!

> Is there a reason why import is only allowed on root modules or is it a just an initial limitation?

This is just the initial implementation. We do intend to expand import blocks within modules in future versions and are exploring a few ideas right now.

If you don't mind, it would be great hear more specific use cases and what you hope to achieve with having import{} within modules.

**Comment by omarismail** [comment]
@ilmax Thanks for providing this info! I'd love to spend time chatting. If you are open to it, please email me at oismail@hashicorp.com and we can schedule a time for a live chat to dig into this!

**Comment by ilmax** [comment]
Email sent!

**Comment by ncrothe** [comment]
Absolutely support the need for this. We're busy importing existing resources from multiple non-Terraform environments into corresponding Terraform states and need a way to say "If we're running this for environment X, please import using this ID".

**Comment by cmlccie** [comment]
This ☝️ would be helpful in modules to address singletons with IDs that follow known patterns. For example, consider AWS [service-linked roles](https://registry.terraform.io/providers/hashicorp/aws/5.4.0/docs/resources/iam_service_linked_role) like the OpenSearch/ElasticSearch service-linked role, which will always have the same _account-specific_ ID. If your module needs this this account-level singleton deployed, there is no good way to ensure the role is deployed and deployed only once.

With ID interpolation for import blocks, this could address singleton issues in modules where the IDs follow known patterns by allowing the module to import the existing resource (if present) and otherwise create the missing resource (if absent) - declaratively stating that the singleton resource must exist:

```hcl
resource "aws_iam_service_linked_role" "elasticsearch" {
  aws_service_name = "es.amazonaws.com"
}

data "aws_caller_identity" "current" {}

import {
    to = aws_iam_service_linked_role.elasticsearch
    id = "arn:aws:iam::${data.aws_caller_identity.current.account_id}:role/aws-service-role/es.amazonaws.com/AWSServiceRoleForAmazonElasticsearchService"
}

**Comment by MoritzHorch** [comment]
For anyone who cant wait: We currently use a hacky workaround in our CI/CD to handle multiple environments / stages. 

One could configure it like:
```
import {
  to = my_terraform_ressource.name
  id = "project/STAGE/the_ressource_id" # note the STAGE placeholder
}
```

And before running any `terraform` command in your pipeline one could use:
```
sed -i 's/STAGE'"$STAGE_VARIABLE"'/g' myfile.tf # replaces STAGE with the content of $STAGE_VARIABLE in myfile.tf
```

Works flawlessly right now.

**Comment by awwright** [comment]
I think this may be similar to my issue #27366 — being able to compute the Terraform ID of an existing resource

**Comment by slushysnowman** [comment]
Related to this, I would also love to be able to do a `for_each` for an import block.

This would massively simplify complex migrations, currently we need to do a lot of complex scripting outside of Terraform using the `terraform import` command to achieve this

**Comment by garymcleanuk** [comment]
> Related to this, I would also love to be able to do a `for_each` for an import block.
> 
> This would massively simplify complex migrations, currently we need to do a lot of complex scripting outside of Terraform using the `terraform import` command to achieve this

Problem is once you have imported from the TF file, that block is no longer used and can be deleted.
Its there much need in adding a for_each?

**Comment by slushysnowman** [comment]
> > Related to this, I would also love to be able to do a `for_each` for an import block.
> > This would massively simplify complex migrations, currently we need to do a lot of complex scripting outside of Terraform using the `terraform import` command to achieve this
> 
> Problem is once you have imported from the TF file, that block is no longer used and can be deleted. Its there much need in adding a for_each?

You're right in a sense, but maybe I can sketch an example that might highlight why I think it would still be valuable:

Say you're managing 10 kubernetes clusters and are in the process of migrating to use Terraform to configure them.

Each cluster can have varying numbers of namespaces, with different names.

If you're planning to define the namespaces via a `for_each` based on some sort of logic (some values file, or variables or whatever), in this migration scenario it makes life so much easier if you can also define a `for_each` on the import block.

The alternative in the current scenario would be something like defining 10x import.tf files containing all the static import blocks for each cluster, and doing it that way, which is quite a painful process.

**Comment by lightpeak** [comment]
Please implement this feature as it would be super useful:

```
import {
    to = 
    id = "/subscriptions/${var.subscriptionId}/resourceGroups/${var.resourceGroupName}/providers/Microsoft.ApiManagement/service/${var.apimInstanceName}/diagnostics/applicationinsights"
}
```

This way we don't have to use the import command in the CLI or don't have to hardcode these id's in the .tf file.

**Comment by ilmax** [comment]
Looks like the PR implementing this [just landed](https://github.com/hashicorp/terraform/pull/33618) 🎉🎉

**Comment by n1ngu** [comment]
As a continuation for this issue, it'd be nice if import blocks supported for-each!

My use-case would be starting to manage resources that can be fetched with a datasource but are most usually created outside the the terraform workflow.
```
data "something" "this" {
  query_params = ???
}
resource "something" "this" {
  for_each = {
    for record in data.something.this.results :
    record.name : record
  }
  name = each.name
  some_option = foo
}

import {
  for_each = {
    for record in data.something.this.results :
    record.name : record
  }
  to = something.this[each.name]
  id = each.identifier
}
```

Right now I am bound to manually run `terraform import something.this[newname] newidentifier` each time somebody creates a resource.

Does this make sense to anyone else?

**Comment by kmoe** [comment]
Wow, you're all on the ball here. Yes, this is closed by https://github.com/hashicorp/terraform/pull/33618.

@n1ngu, please open a separate issue for `for_each` within the `import` block with your helpful use case.

**Comment by slushysnowman** [comment]
@n1ngu fully agreed, see my earlier comments, if you don't create the separate issue for it I will ;-)

**Comment by kmoe** [comment]
Today's v1.6.0 alpha release supports this feature - interested to hear your feedback here (UX) or in a new issue (bugs, crashes).

**Comment by ilmax** [comment]
@slushysnowman mind you opening another issue for supporting for each since the idea here came from you?

**Comment by slushysnowman** [comment]
I've created #33624 for the request for `for_each` support

**Comment by nwaughachukwuma** [comment]
> Looks like the PR implementing this [just landed](https://github.com/hashicorp/terraform/pull/33618) 🎉🎉

Has it? I still can do variable interpolation

```hcl
import {
  to = google_firestore_database.database
  id = "projects/${var.project}/databases/(default)"
}

Error: Unsuitable value type
│ 
│   on imports.tf line 59, in import:
│   59:   id = "projects/${var.project}/databases/(default)"
│ 
│ Unsuitable value: value must be known
╵
╷
│ Error: Variables not allowed
│ 
│   on imports.tf line 59, in import:
│   59:   id = "projects/${var.project}/databases/(default)"
│ 
│ Variables may not be used here.
╵
```

**Comment by crw** [comment]
@nwaughachukwuma See https://github.com/hashicorp/terraform/issues/33228#issuecomment-1662321315

If you are using 1.6.x and have an issue, please open a new issue following the full new issue template. Thanks!

**Comment by nwaughachukwuma** [comment]
Ok, thanks.
